File: Backend/src/main/java/com/pro/booking/entity/model/Room.java
package com.pro.booking.entity.model;

import com.pro.booking.core.base.BaseEntity;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.util.HashSet;
import java.util.Set;

@EqualsAndHashCode(callSuper = true)
@Entity
@Table(name = "rooms")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Room extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "host_id", nullable = false)
    private User host;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "room_categories",
            joinColumns = @JoinColumn(name = "room_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private Set<Category> categories = new HashSet<>();


    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "room_commons",
            joinColumns = @JoinColumn(name = "room_id"),
            inverseJoinColumns = @JoinColumn(name = "common_id")
    )
    private Set<Category> commons = new HashSet<>();


    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "room_tags",
            joinColumns = @JoinColumn(name = "room_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<Category> tags = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "documents",
            joinColumns = @JoinColumn(name = "room_id"),
            inverseJoinColumns = @JoinColumn(name = "Document_id")
    )
    private Set<Document> documents = new HashSet<>();



    @Column(nullable = false)
    private String title;

    @Lob
    private String subTitle;

    @Lob
    private String description;

    @Column(name = "price_per_night", nullable = false, precision = 15, scale = 2)
    private BigDecimal pricePerNight;

    @Column(name = "cleaning_fee", precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal cleaningFee = BigDecimal.ZERO;

    @Column(name = "service_fee", precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal serviceFee = BigDecimal.ZERO;

    @Column(name = "max_guests")
    @Builder.Default
    private Long maxGuests = 1L;

    @Column(name = "room_count")
    @Builder.Default
    private Long roomCount = 1L;

    @Column(name = "bathroom_count")
    @Builder.Default
    private Long bathroomCount = 1L;
}

File: staynia/lib/data/entity/model/room.dart
import 'package:staynia/data/entity/model/category.dart';
import 'package:staynia/data/entity/model/document.dart';

class Room {
  int? status;
  String? actionKey;
  String? createdDate;
  String? lastUpdatedDate;
  int? id;
  List<Category>? categories;
  List<Category>? commons;
  List<Category>? tags;
  List<Document>? documents;
  String? title;
  String? subTitle;
  String? description;
  double? pricePerNight;
  double? cleaningFee;
  double? serviceFee;
  int? maxGuests;
  int? roomCount;
  int? bathroomCount;

  Room({
    this.status,
    this.actionKey,
    this.id,
    this.categories,
    this.commons,
    this.documents,
    this.title,
    this.description,
    this.tags,
    this.subTitle,
    this.pricePerNight,
    this.cleaningFee,
    this.serviceFee,
    this.maxGuests,
    this.roomCount,
    this.createdDate,
    this.lastUpdatedDate,
    this.bathroomCount,
  });

  factory Room.fromJson(Map<String, dynamic> json) {
    return Room(
      status: json['status'],
      actionKey: json['actionKey'],
      id: json['id'],
      categories: (json['categories'] as List?)
          ?.map((e) => Category.fromJson(e))
          .toList(),
      commons: (json['commons'] as List?)
          ?.map((e) => Category.fromJson(e))
          .toList(),
      tags: (json['tags'] as List?)
          ?.map((e) => Category.fromJson(e))
          .toList(),
      documents: (json['documents'] as List?)
          ?.map((e) => Document.fromJson(e))
          .toList(),
      title: json['title'],
      createdDate: json['createdDate'],
      lastUpdatedDate: json['lastUpdatedDate'],
      subTitle: json['subTitle'],
      description: json['description'],
      pricePerNight: (json['pricePerNight'] as num?)?.toDouble(),
      cleaningFee: (json['cleaningFee'] as num?)?.toDouble(),
      serviceFee: (json['serviceFee'] as num?)?.toDouble(),
      maxGuests: json['maxGuests'],
      roomCount: json['roomCount'],
      bathroomCount: json['bathroomCount'],
    );
  }

  Map<String, dynamic> toJson() => {
    'status': status,
    'actionKey': actionKey,
    'id': id,
    'title': title,
    'description': description,
    'pricePerNight': pricePerNight,
    'cleaningFee': cleaningFee,
    'serviceFee': serviceFee,
    'maxGuests': maxGuests,
    'roomCount': roomCount,
    'bathroomCount': bathroomCount,
  };
}
